hazelcast/src/main/java/com/hazelcast/nio/NodeIOService.java

```diff
@@ -332,5 +382,39 @@ private void transformPortDefinitionsToPorts(Collection<String> portDefinitions,
         return networkConfig.getOutboundPortDefinitions() == null
                 ? Collections.<String>emptySet() : networkConfig.getOutboundPortDefinitions();
     }
+
+    private class ThroughputMonitor extends Thread {
```

Comment:
I would pull this out into its own class. This class already has a huge amount of functionality; no need to add even more to it. Also the packet counters can be simplified to volatile longs. It is a bit more work, but you will remove contention from the heart of the system.